contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c

```diff
@@ -374,6 +374,228 @@ tsql_check_param_readonly(const char *paramname, TypeName *typename, bool readon
+
+        }
+
+        jsonWithParams = lappend(jsonWithParams, makeStringConst(format_cols->data, -1));
+        rf->coldeflist = lappend(rf->coldeflist, n);
+    }
+
+    fc = makeFuncCall(TsqlSystemFuncName("openjson_with"), jsonWithParams, COERCE_EXPLICIT_CALL, -1);
+    rf->functions = list_make1(list_make2(fc, NULL));
+    rf->alias = alias;
+    return (Node*) rf;
+}
+
+/*
+* Create a column definition node for the given column name and type
+* If the column type is a character type, we need to change the underlying typmod
+*/
+Node *
+createOpenJsonWithColDef(char* elemName, TypeName* elemType)
+{
+    ColumnDef *n = makeNode(ColumnDef);
+    char* typeNameString = TypeNameToString(elemType);
+    n->colname = elemName;
+    if(isCharType(typeNameString))
+    {
+        n->typeName = setCharTypmodForOpenjson(elemType);
+    }
+    else
+    {
+        n->typeName = elemType;
+    }
+    n->inhcount = 0;
+    n->is_local = true;
+    n->is_not_null = false;
+    n->is_from_type = false;
+    n->storage = 0;
+    n->raw_default = NULL;
+    n->cooked_default = NULL;
+    n->collOid = InvalidOid;
+    n->constraints = NIL;
+    n->location = -1;
+    return (Node*) n;
+}
+
+TypeName *
+setCharTypmodForOpenjson(TypeName *t)
+{
+    int curTMod = getElemTypMod(t);
+    List *tmods = (List*) t->typmods;
+    if(tmods == NULL)
+    {
+        // Default value when no typmod is provided is 1
+        t->typmods = list_make1(makeIntConst(1, -1));
+        return t;
+    }
+    else if(curTMod == TSQLMaxTypmod)
+    {
+        // TSQLMaxTypmod is represented as -8000 so we need to change to
+        // the actual max value of 4000
+        t->typmods = list_make1(makeIntConst(4000, -1));
+        return t;
+    }
+    else
+    {
+        return t;
+    }
+}
+
+bool isCharType(char* typenameStr)
+{
+    if(pg_strcasecmp(typenameStr, "char") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "nchar") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "varchar") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "pg_catalog.char") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "pg_catalog.nchar") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "pg_catalog.varchar") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "sys.char") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "sys.nchar") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "sys.varchar") == 0)
+    {
+        return true;
+    }
+    else if(isNVarCharType(typenameStr))
+    {
+        return true;
+    }
+    return false;
+}
+
+bool isNVarCharType(char* typenameStr)
+{
+    if(pg_strcasecmp(typenameStr, "nvarchar") == 0)
+    {
+        return true;
+    }
+    else if(pg_strcasecmp(typenameStr, "pg_catalog.nvarchar") == 0)
```

Comment:
1. There is no `pg_catalog.nvarchar` type.
2. Please add tests for all cases in `isNVarCharType()` and `isCharType()`.